crates/memory/src/memory.rs

```diff
@@ -359,3 +359,72 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
         borrows.unborrow_mut(self.handle);
     }
 }
+
+pub struct GuestArray<'a, T>
+where
+    T: GuestType,
+{
+    ptr: GuestPtr<'a, T>,
+    num_elems: u32,
+}
+
+impl<'a, T> GuestArray<'a, T>
+where
+    T: GuestTypeCopy,
+{
+    pub fn as_ref(&'a self) -> Result<&'a [T], GuestError> {
+        // First, validate
+        for i in 0..self.num_elems {
+            let ptr = self.ptr.elem(i as i32)?;
+            T::validate(&ptr)?;
+        }
+        // Then, extract as slice
+        Ok(unsafe {
+            std::slice::from_raw_parts(self.ptr.as_raw() as *const T, self.num_elems as usize)
+        })
```

Comment:
I think this is slightly better now compared to what it was in the previous commit d910c0e. I hope I'm right in my thinking here that by firstly validating that all (expected) pointers are valid, we can then safely extract the slice `&'a [T]` utilising `std::slice::from_raw_parts` with a raw pointer cast.

Reply:
I believe this is correct as well.